### ISSUE TITLE
Get available dates not working anymore

### ISSUE THREAD

**Issue by jyepesr1** [dvalbuena1/visa_rescheduler_aws#10]
I can confirm the method to get the available dates via the JSON is not working anymore, this is a proposal to select the date "manually" from the date picker https://github.com/uxDaniel/visa_rescheduler/pull/39

I would like to know your thoughts @dvalbuena1

**Comment by dvalbuena1** [comment]
I think you are absolutely right, I always thought that when checking the enpoint it returned 404 because no date was found. But it really seems that it is not so easy to get that information anymore. I will check if there is a way to not do it manually, if I can't find it the manual approach could be useful and I will make a commit with the change.

**Comment by jyepesr1** [comment]
@dvalbuena1 I also saw this approach https://github.com/ahmetgunduz/visa_rescheduler/blob/main/visa.py#L159, If this works would be a lot easier.

**Comment by dvalbuena1** [comment]
I just did a commit  956b98d420af580e08022f00786f20bc7d03728f with a similar approach and it seems to be working. Please check.

**Comment by jyepesr1** [comment]
hey, @dvalbuena1 I'll take a look, should we do the same approach for `get_time` and ASC `date` and `time` methods as well?

something like...

```python
def get_time(self, date):
    time_url = TIME_URL.format(date=date)
    headers = {
        "Accept": "application/json, text/javascript, */*; q=0.01",
        "X-Requested-With": "XMLHttpRequest",
        "User-Agent": self.driver.execute_script("return navigator.userAgent;"),
        "Referer": APPOINTMENT_URL,
        "Cookie": "_yatri_session="
        + self.driver.get_cookie("_yatri_session")["value"],
    }
    r = requests.get(time_url, headers=headers)
    data = r.json()
    time = data.get("available_times")[-1]
    print(f"Got time successfully! {date} {time}")
```

**Comment by dvalbuena1** [comment]
Sorry for the late response.
I just did a new commit 3d25965 that should fix the way it retrieves the dates and the times, even for ASC cases.
Let me know if everything is working now.

**Comment by Ravi-099** [comment]
It doesn't seem to work anymore. It's crashing when there are no dates available and shows INFO:send_notification - Sending notification: HELP! Crashed.

**Comment by dvalbuena1** [comment]
For me it has been working, if you can get the AWS logs with those errors it would be easier to detect the error.

**Comment by Ravi-099** [comment]
I am using it locally.



```
DevTools listening on ws://127.0.0.1:54611/devtools/browser/d5198261-4d8c-4c10-852f-4a003d94a3e6
INFO:main - ---START--- : 2023-08-13 20:55:45.757043
[12888:13740:0813/205548.230:ERROR:gpu_disk_cache.cc(216)] Failed to create blob cache entry: -2
INFO:login - Login start...
INFO:login -    click bounce
INFO:do_login_action -  input email
INFO:do_login_action -  input pwd
INFO:do_login_action -  click privacy
INFO:do_login_action -  commit
INFO:send_notification - Sending notification: HELP! Crashed.
ERROR:main - Message:
Stacktrace:
Backtrace:
        GetHandleVerifier [0x011C2B63+47795]
        (No symbol) [0x011567E1]
        (No symbol) [0x0105C7ED]
        (No symbol) [0x01089AC8]
        (No symbol) [0x01089B8B]
        (No symbol) [0x010B8A82]
        (No symbol) [0x010A4D14]
        (No symbol) [0x010B71DA]
        (No symbol) [0x010A4AC6]
        (No symbol) [0x010813D7]
        (No symbol) [0x0108255D]
        GetHandleVerifier [0x01416F89+2490073]
        GetHandleVerifier [0x0145C361+2773681]
        GetHandleVerifier [0x01456211+2748769]
        GetHandleVerifier [0x012416B0+566784]
        (No symbol) [0x0116019A]
        (No symbol) [0x0115C418]
        (No symbol) [0x0115C4FB]
        (No symbol) [0x0114F317]
        BaseThreadInitThunk [0x755E00C9+25]
        RtlGetAppContainerNamedObjectPath [0x77757B4E+286]
        RtlGetAppContainerNamedObjectPath [0x77757B1E+238]


```

**Comment by jyepesr1** [comment]
@dvalbuena1 For me it's working in AWS, the issue is the Embassy still has the re-schedule disabled at least in Colombia, so I never get a chance to test the full workflow, it's always returning []

**Comment by jameszhan02** [comment]
> @dvalbuena1 For me it's working in AWS, the issue is the Embassy still has the re-schedule disabled at least in Colombia, so I never get a chance to test the full workflow, it's always returning []

I guess is just not working any more, Since when they updated the site

**Comment by jameszhan02** [comment]
> > @dvalbuena1 For me it's working in AWS, the issue is the Embassy still has the re-schedule disabled at least in Colombia, so I never get a chance to test the full workflow, it's always returning []
> 
> I guess is just not working any more, Since when they updated the site

I mean imagen those people who work for the US visa to build this site, just bowering around the github, fixing 'bugs' by the code posted  LOL

**Comment by js-segura-uniandes** [comment]
> For me it has been working, if you can get the AWS logs with those errors it would be easier to detect the error.


For me it works. But it never leaves here:

![image](https://github.com/dvalbuena1/visa_rescheduler_aws/assets/78027564/b6537073-351b-49db-aad7-0319f6f2ab19)

Do you know if I could be banned for trying so hard? @dvalbuena1

**Comment by jameszhan02** [comment]
> > For me it has been working, if you can get the AWS logs with those errors it would be easier to detect the error.
> 
> For me it works. But it never leaves here:
> 
> ![image](https://user-images.githubusercontent.com/78027564/277066886-b6537073-351b-49db-aad7-0319f6f2ab19.png)
> 
> Do you know if I could be banned for trying so hard? @dvalbuena1

1.  if u do each request with 3 mins  gap, you will likely to get banned in a day. but is not a big deal, just need to w8 for a half day then you can start request again

**Comment by jonsnowljs** [comment]
> > > For me it has been working, if you can get the AWS logs with those errors it would be easier to detect the error.
> > 
> > 
> > For me it works. But it never leaves here:
> > ![image](https://user-images.githubusercontent.com/78027564/277066886-b6537073-351b-49db-aad7-0319f6f2ab19.png)
> > Do you know if I could be banned for trying so hard? @dvalbuena1
> 
> 1. if u do each request with 3 mins  gap, you will likely to get banned in a day. but is not a big deal, just need to w8 for a half day then you can start request again

May I know what's your gap. I set the gap as 160s and got banned after 1 hour. Create multiple DS160 to check the schedule might be helpful.

**Comment by js-segura-uniandes** [comment]
@dvalbuena1 @jameszhan02 @jonsnowljs 

Have they managed to schedule? It's been running for three days and there are never any dates available.

**Comment by jameszhan02** [comment]
> @dvalbuena1 @jameszhan02 @jonsnowljs 
> 
> Have they managed to schedule? It's been running for three days and there are never any dates available.

i had successfully book one in Aug for myself , but currently I try to do this for my friend. I did modified bit code for some changes they made for this site. I run for 2 days only 2 times have available dates response back from api and they all in 2025, so I guess it's  really have no date available currently for the visa

**Comment by Ravi-099** [comment]
> > @dvalbuena1 @jameszhan02 @jonsnowljs
> > Have they managed to schedule? It's been running for three days and there are never any dates available.
> 
> i had successfully book one in Aug for myself , but currently I try to do this for my friend. I did modified bit code for some changes they made for this site. I run for 2 days only 2 times have available dates response back from api and they all in 2025, so I guess it's really have no date available currently for the visa

@jameszhan02 Could you please tell me what changes you have made to the code? I have tried removing the UTF line but it doesn't work.

**Comment by Ravi-099** [comment]
I was able to get around this error by commenting on it. However, it doesn't return r.status_code == 200: so the notification doesn't update. 

`        data = {
            **# "utf8": self.driver.find_element(by=By.NAME, value='utf8').get_attribute('value'),**
            "authenticity_token": self.driver.find_element(by=By.NAME, value='authenticity_token').get_attribute(
                'value'),`

**Comment by jameszhan02** [comment]
> > > @dvalbuena1 @jameszhan02 @jonsnowljs
> > > Have they managed to schedule? It's been running for three days and there are never any dates available.
> > 
> > 
> > i had successfully book one in Aug for myself , but currently I try to do this for my friend. I did modified bit code for some changes they made for this site. I run for 2 days only 2 times have available dates response back from api and they all in 2025, so I guess it's really have no date available currently for the visa
> 
> @jameszhan02 Could you please tell me what changes you have made to the code? I have tried removing the UTF line but it doesn't work.

https://github.com/jameszhan02/visa_rescheduler/tree/FM

**Comment by dvalbuena1** [comment]
Sorry for the late reply. 
In this commit 3d25965b53bcf892ff986c432c242cb8b64fa5cc I modified the way in which the dates were obtained, but do not worry if in the logs you see no appointments available, it is because there really are none. Since there have been very few occasions where I have been able to find some appointments. The problem is that being so few availability it is very complicated now to test the code and validate that the rescheduling of the appointment is working correctly.

**Comment by dvalbuena1** [comment]
> I was able to get around this error by commenting on it. However, it doesn't return r.status_code == 200: so the notification doesn't update.
> 
> ` data = { **# "utf8": self.driver.find_element(by=By.NAME, value='utf8').get_attribute('value'),** "authenticity_token": self.driver.find_element(by=By.NAME, value='authenticity_token').get_attribute( 'value'),`

@Ravi-099 Are you sure the error is in the reschedule method? Because you should have seen this log before the error.

` logger.info(f"New date: {date} {date_time}")`

**Comment by Ravi-099** [comment]
> > I was able to get around this error by commenting on it. However, it doesn't return r.status_code == 200: so the notification doesn't update.
> > ` data = { **# "utf8": self.driver.find_element(by=By.NAME, value='utf8').get_attribute('value'),** "authenticity_token": self.driver.find_element(by=By.NAME, value='authenticity_token').get_attribute( 'value'),`
> 
> @Ravi-099 Are you sure the error is in the reschedule method? Because you should have seen this log before the error.
> 
> ` logger.info(f"New date: {date} {date_time}")`

This was the error message when it tried to reschedule. Last night I was able to run this code by using the above workaround. 

`````
DevTools listening on ws://127.0.0.1:51783/devtools/browser/7d8ced59-d43f-42d8-adbc-a60b4fa85e4c
INFO:main - ---START--- : 2023-10-20 23:35:19.102885
INFO:login - Login start...
INFO:login -    click bounce
INFO:do_login_action -  input email
INFO:do_login_action -  input pwd
INFO:do_login_action -  click privacy
INFO:do_login_action -  commit
INFO:do_login_action -  login successful!
INFO:print_dates - Available dates:
INFO:print_dates - 2024-10-21    business_day: True
INFO:print_dates - 2024-11-18    business_day: True
INFO:print_dates - 2025-04-07    business_day: True
INFO:print_dates - 2025-07-14    business_day: True
INFO:print_dates - 2025-10-16    business_day: True
INFO:get_available_date - Checking for an earlier date:
INFO:is_earlier - Is 2024-10-31 00:00:00 > 2024-10-21 00:00:00: True
INFO:get_time - Got time successfully! 2024-10-21 08:30
INFO:main - New date: 2024-10-21 08:30
INFO:reschedule - Starting Reschedule (2024-10-21)
INFO:reschedule -       single applicant
INFO:send_notification - Sending notification: HELP! Crashed.
ERROR:main - Message: no such element: Unable to locate element: {"method":"css selector","selector":"[name="utf8"]"}
`````

I believe it shows this error message because of an element change in the website. This was the website code when the dates were available. 

`````
<form id="appointment-form" novalidate="novalidate" class="formtastic appointments" action="/en-ca/niv/schedule/50415214/appointment" accept-charset="UTF-8" method="post"><input type="hidden" name="authenticity_token" value="8IxtoecmFBjCKMTX0qIX3H0mvs+W1Bj4Y/gjnYMdqQ/x36e7jZlt/N5KBskdDN80gRVB/pCWYcdEU6N9CgQ2bA==" autocomplete="off" /><input type="hidden" name="confirmed_limit_message" id="confirmed_limit_message" value="1" autocomplete="off" />
`````

**Comment by dvalbuena1** [comment]
After some research the fix was made in the last commit ea5ad23

Also, a bug in case of multiple groups in the same account has been fixed.

**And to keep in mind, the lambda function must be completely deleted `sls delete` and redeployed, because an improvement in the deployment was made.**

**Comment by Eljeff13** [comment]
chicos, me presenta este error al correr la función en AWS, alguien podria ayudar?
me imagino que es de configuración en el dockerfile. 

![Screenshot_1](https://github.com/dvalbuena1/visa_rescheduler_aws/assets/39741767/4d5d6390-3319-4f05-9349-a741e440ab02)
